### ISSUE TITLE
Bug in muon FSR correction with offsets

### ISSUE THREAD

**Issue by jpata** [jpata/hepaccelerate-cms#56]
I was setting up some tests (PR #55) on an unskimmed 2016 sync file and I noticed an issue with the muon FSR correction code on [hmumu_utils.py#L1228](https://github.com/jpata/hepaccelerate-cms/blob/6cb6d235df9bbd1c03f2c3314a5a9cf0890d4e52/tests/hmm/hmumu_utils.py#L1228). You can see the automatic test failing [here](https://travis-ci.com/jpata/hepaccelerate-cms/builds/134634776).

It's not correct to assume that one can use `offsets[0:-1]` to index into the underlying data array. In particular, if e.g. the last event has 0 muons, the offsets will be `[ ... 181229 181231 181231]`, where the muon data array has length 181231. This did not show up in testing as we apply a skim on the HLT bit, which removes such events. I think that there should be no physics effect from this bug if there is no error, but I'm not 100% sure. In this case, the error arises because we are doing manual offset manipulation. I think offset manipulation would be best left to either:
1) an explicit Numba loop, which will throw an error/assert and can be checked more easily
2) use awkward-array library (we so far use it minimally as it does not support GPUs yet)

One can test this with the following script: 
```python
import uproot

tt = uproot.open("data/myNanoProdMc2016_NANO.root").get("Events")
mu_pt = tt.array("Muon_pt")
print(mu_pt.content.shape)
print(mu_pt.offsets)

for iev in range(len(mu_pt.offsets) - 1):
    imu1 = mu_pt.offsets[iev]
    imu2 = mu_pt.offsets[iev+1]
    print(f"Event {iev} has {imu2-imu1} muons (offsets {imu1}:{imu2})")
    for imu in range(imu1, imu2):
        print(imu, mu_pt.content[imu])

mu_pt.content[mu_pt.offsets[0:-1]]
```

Test file: [myNanoProdMc2016_NANO.root](https://jpata.web.cern.ch/jpata/hmm/test_files/myNanoProdMc2016_NANO.root)

@kondratyevd if you have time next week, could you take a look at this? To me, a CPU-only numba function such as the following:
```python
#Corrects the muon momentum and isolation, if a matched FSR photon with dR<0.4 is found
@numba.njit
def correct_muon_with_fsr(
        muons_offsets, fsr_offsets,
        muons_pt, muons_eta, muons_phi, muons_mass, muons_iso, muons_fsrIndex
        fsr_pt, fsr_eta, fsr_phi, fsr_mass,
        out_muons_pt, out_muons_eta, out_muons_phi, out_muons_mass, out_muons_iso
    ):
    for iev in numba.prange(len(muons_offsets) - 1):

        #loop over muons in event
        mu1 = muons_offsets[iev]
        mu2 = muons_offsets[iev + 1]

        for imu in range(mu1, mu2):

            #relative FSR index in the event 
            fsr_idx_relative = muons_fsrIndex[imu]
            if fsr_idx_relative >= 0:
                #absolute index in the full FSR vector for all events
                ifsr = fsr_offsets[iev] + fsr_idx_relative

                this_mu_pt = muons_pt[imu]
                ... (and others)

                this_fsr_pt = fsr_pt[ifsr]
                ... (and others)

                #compute dR between muon and FSR, check <0.4
                #compute and set corrected isolation
                #compute and set corrected momentum
```
would perhaps be easier to follow and debug. What do you think? I'm also open to alternatives...

**Comment by kondratyevd** [comment]
Hi @jpata,
I'll look into it on Monday. I think a loop in numba should work fine, I was going to implement it anyway after we have used a similar approach with jet PU ID.

**Comment by jpata** [comment]
solved in #59 - thanks!
